sdks/go/examples/native_wordcap/nativepubsubio/native.go

```diff
@@ -0,0 +1,247 @@
+	client            *pubsub.Client
+	processedMessages []*pubsub.Message
+}
+
+// newPubSubRead inserts an unbounded read from a PubSub topic into the pipeline. If an existing subscription
+// is provided, the DoFn will read using that subscription; otherwise, a new subscription to the topic
+// will be created using the provided subscription name.
+func newPubSubRead(ctx context.Context, projectID, topic, subscription string) (*pubSubRead, error) {
+	if topic == "" {
+		return nil, errors.New("please provide either a topic to read from")
+	}
+	client, err := pubsub.NewClient(ctx, projectID)
+	if err != nil {
+		return nil, err
+	}
+
+	top := client.Topic(topic)
+	if ok, err := top.Exists(ctx); !ok || err != nil {
+		return nil, fmt.Errorf("failed to get topic; exists: %v, error: %v", ok, err)
+	}
+	sub, err := pubsubx.EnsureSubscription(ctx, client, topic, subscription)
+	if err != nil {
+		return nil, err
+	}
+	return &pubSubRead{ProjectID: projectID, Subscription: sub.ID()}, nil
+}
+
+// CreateInitialRestriction() establishes the PubSub subscription ID as the
+// initial restriction
+func (r *pubSubRead) CreateInitialRestriction(_ []byte) string {
+	return r.Subscription
+}
+
+// CreateTracker wraps the PubSub subscription ID in a StaticRTracker
+// and applies a mutex via LockRTracker.
+func (r *pubSubRead) CreateTracker(rest string) *sdf.LockRTracker {
+	return sdf.NewLockRTracker(NewSubscriptionRTracker(rest))
+}
+
+// RestrictionSize always returns 1.0, as the restriction is always 1 subscription.
+func (r *pubSubRead) RestrictionSize(_ []byte, rest string) float64 {
+	return 1.0
+}
+
+// SplitRestriction is a no-op as the restriction cannot be split.
+func (r *pubSubRead) SplitRestriction(_ []byte, rest string) []string {
+	return []string{rest}
+}
+
+// Setup initializes a PubSub client if one has not been created already
+func (r *pubSubRead) Setup(ctx context.Context) error {
+	if r.client == nil {
+		client, err := pubsub.NewClient(ctx, r.ProjectID)
+		if err != nil {
+			return err
+		}
+		r.client = client
+	}
+	return nil
+}
+
+var messageTimeout time.Duration = 5 * time.Second
+
+// ProcessElement initializes a PubSub client if one has not been created already, reads from the PubSub subscription,
+// and emits elements as it reads them. If no messages are available, the DoFn will schedule itself to resume processing
+// later. If polling the subscription returns an error, the error will be logged and the DoFn will not reschedule itself.
+func (r *pubSubRead) ProcessElement(ctx context.Context, bf beam.BundleFinalization, rt *sdf.LockRTracker, _ []byte, emit func(beam.EventTime, []byte)) (sdf.ProcessContinuation, error) {
+	// Register finalization callback
+	bf.RegisterCallback(5*time.Minute, func() error {
+		for _, m := range r.processedMessages {
+			m.Ack()
+		}
+		r.processedMessages = nil
+		return nil
+	})
+
+	for {
+		ok := rt.TryClaim(r.Subscription)
+		if !ok {
+			return sdf.ResumeProcessingIn(5 * time.Second), nil
+		}
```

Comment:
As written, we're sort of prevented from having the runner force a checkpoint (only allowing DoFn checkpoints, but given we don't currently split, this is probably fine for now.  

The risk is a higher throughput pubsub, which will prevent this code from checkpointing, which will prevent downstream aggregations.

Reply:
For a short example I'm not going to sweat it too much, we can revisit how the runner would split this later.